File: Graph/1376. Time Needed to Inform All Employees/1376. Time Needed to Inform All Employees.java
class Solution {
    // creat graph and dfs get path sum
    int rets = 0;
    public int numOfMinutes(int n, int headID, int[] manager, int[] informTime) {
        Map<Integer, List<Integer>> map = new HashMap<>();
        for (int i = 0; i < manager.length; i++) {
            if (manager[i] != -1) {
                map.computeIfAbsent(manager[i], k -> new ArrayList<>()).add(i);
            }
        }
        dfs(map, headID, 0, informTime);
        return rets;
    }
    private void dfs(Map<Integer, List<Integer>> map, int pos, int time, int[] informTime) {
        if (!map.containsKey(pos)) {
            rets = Math.max(rets, time);
            return;
        }
        
        int nexttime = time+informTime[pos];
        for (int next : map.get(pos)) {
            dfs(map, next, nexttime, informTime);
        }
    }
}

File: Graph/1376. Time Needed to Inform All Employees/1376. Time Needed to Inform All Employees.cpp
class Solution {
public:
    int numOfMinutes(int n, int headID, vector<int>& manager, vector<int>& informTime) {
        int rets = 0;
        for (int i = 0; i < n; i++) {
            rets = max(rets, dfs(i, manager, informTime));
        }
        return rets;
    }
    int dfs(int pos, vector<int>& manager, vector<int>& informTime) {
        if (manager[pos] != -1) {
            informTime[pos] += dfs(manager[pos], manager, informTime);
            manager[pos] = -1;
        }
        return informTime[pos];
    }
};
